### ISSUE TITLE
Add some array features

### ISSUE THREAD

**Issue by luisherranz** [xamfoo/reactive-obj#5]
Congratulations @xamfoo, I think this package has an incredible potential! :)

I've been studying the Flux architecture lately and I really like the idea of having all the app state in a single global object. It looks like Facebook is starting to use [baobab](https://github.com/Yomguithereal/baobab) with ReactJS and Optimizely is already using something similar in [NuclearJS](https://github.com/optimizely/nuclear-js).

I like baobab but most of the things it can do are already covered by your excellent package. I think the only thing I'm missing right now are some array features, like push, unshift or splice.
https://github.com/Yomguithereal/baobab#cursor-level

I'd love to know your thoughts about baobab as well.

**Comment by xamfoo** [comment]
@LuisHerranz you are very kind. Thank you for the suggestion :)

baobab seems interesting. Some observations:
- Facets look similar to getters in NuclearJS. I'd like to see how this can work for reactive data sources.
- I would prefer synchronous updating to asynchronous batch updating in baobab. One could set `asynchronous` option to false but seems like this would affect its performance.
- History and updating commands are nice but doesn't seem necessary to me.
- Cursors look useful, particularly in this case where we want to apply array methods to a path.

Array methods can look like:

``` javascript
var x = new ReactiveObj({y: []});
var y = x.select('y');
y.push('foo');
y.push('bar');
y.pop(); // 'bar';
x.get(); // {y: ['foo']}
```

This is certainly open for discussion

**Comment by luisherranz** [comment]
- Yeah, cursors would be a great addition to simplify the use of deep ReactiveObjs.
- [Tracker is asynchronous](http://docs.meteor.com/#/full/tracker_flush), so updates in Meteor are async by default anyway.
- History and undo capabilities would be great for some applications, but probably not needed everywhere. Those come for free when using immutable data, something maybe out of the scope here.
- Array methods look great! Maybe you can use them without cursors as well:

``` javascript
var x = new ReactiveObj({ a: { b: [] } });
x.push(['a', 'b'], 'foo');
x.push(['a', 'b'], 'bar');
x.get(); // { a: { b: [ 'foo', 'bar' ] } }
```

**Comment by xamfoo** [comment]
Good point about using them without cursors. I'd probably add `push`, `pop`, `reverse`, `shift`, `sort`, `splice` and `unshift`.

Let me try to clarify my earlier comment about updating.

Note **>>** means synchronous, **>>async>>** means asynchronous.
1. When we want to read a new value in baobab:
   - Set value **>>async>>** Update value **>>** Trigger callbacks **>>** Read new value
2. while for `ReactiveDict` or `ReactiveObj`:
   - Set value **>>** Update value **>>** Read new value **>>async>>** Run invalidated computations

I prefer the second case since we don't need to wait for a flush cycle or callbacks to read the value.

I'm interested in the way you are using the global app state. Do you store subscription handlers and Mongo cursors in it?

**Comment by luisherranz** [comment]
Yeah, I think in the end what you want to avoid is refreshing the UI more than needed. I agree this approach is better.

I'm still investigating on how to use it best. The idea is that instead of having app state stored in individual reactive variables inside each Flux Store, you use a global object like ReactiveObj.

On the other side (the templates) you have an API to retrieve those values. Something like this:

``` javascript
// instead of this:
Template.Something.helpers({
  something: function(){
    return SomeStore.getSomeState();
  }
});
// the template doesn't need to know about "SomeStore" and can go with:
Template.Something.helpers({
  something: function(){
    return AppState.getSomeState();
  }
});
```

The change is subtle, but it decouples Templates from Stores.

That "AppState" object can have an API which needs to be declared:

``` javascript
AppState.getSomeState = function() {
   return state.get([ 'state', 'some']);
}
```

Or can be something like ReactObj itself:

``` javascript
Template.Something.helpers({
  something: function(){
    return AppState.get([ 'state', 'some']);
  }
});
```

This can be used for session app state. You need collections for persistent app state of course.

**Comment by xamfoo** [comment]
Array methods implemented in v0.5.0 7831e7524895a3e2932e2017114728c5a9826158

Let me know if it is good for you :)

**Comment by luisherranz** [comment]
Lovely, thanks!

I'll give it a try.
